### PR TITLE
Add downstream eval in training loop

### PR DESCRIPTION
**Pull request by yawenzzzz**
- Added a callback to run downstream evaluation tasks (can be a list of tasks) during training. It currently supports classification but can be extended for other tasks
- Fixed the train/loss name issue
- Some clean up

### CODE REVIEW

File: helios/data/normalize.py [hunk 1]
```diff
@@ -18,8 +18,8 @@ class Strategy(Enum):
 
 # Normalization strategy for each modality
 NORMALIZE_STRATEGY = {
-    Modality.SENTINEL1: Strategy.PREDEFINED,
-    Modality.SENTINEL2: Strategy.PREDEFINED,
+    Modality.SENTINEL1: Strategy.COMPUTED,
```
Comment (Hgherzog) [inline helios/data/normalize.py]: Why are we changing this? Should this be a constant or configruable for a specific run?

File: helios/nn/flexihelios.py [hunk 2]
```diff
@@ -351,7 +351,8 @@ def _apply_encodings_per_modality(
         Returns:
             Tensor with encodings applied based on modality type
         """
-        logger.info(f"Applying encodings to modality {modality}")
```
Comment (Hgherzog) [inline helios/nn/flexihelios.py:354]: nit: make logger debug instead

File: helios/train/callbacks/evaluator_callback.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                f"Downstream evaluator {self.name} {METRIC_NAME} score: {val_result}"
+            )
+            return val_result
+        except Exception as e:
+            logger.error(f"Error during evaluation: {e}")
+            return 0
+
+
+@dataclass
+class DownstreamEvaluatorCallback(Callback):
```
Comment (Hgherzog) [inline helios/train/callbacks/evaluator_callback.py:88]: Suggestion: add a test for this

File: helios/train/loss.py [hunk 4]
```diff
@@ -21,6 +21,10 @@
 class Loss(ABC):
     """Abstract base class for loss functions."""
 
+    def __init__(self, name: str) -> None:
```
Comment (Hgherzog) [inline helios/train/loss.py]: Suggestion: make this a class variable as all instances of a specific type of loss function should have the same name.

File: helios/train/train_module/latent_mim.py [hunk 5]
```diff
@@ -184,7 +178,7 @@ def train_batch(self, batch: HeliosSample, dry_run: bool = False) -> None:
         decoded, loss = self.model_forward(masked_batch, patch_size)
 
         self.trainer.record_metric(
-            TRAIN_PATCH_DISC_LOSS_METRIC,
+            f"train/{self.base_loss.name}",  # get loss name
```
Comment (Hgherzog) [inline helios/train/train_module/latent_mim.py]: suggestion: create a static method for train related metrics

File: helios/data/normalize.py (hunk 1 above)
Comment (gabrieltseng) [reply helios/data/normalize.py]: I think maybe configurable from a config? But in general I would suggest preferring the computed strategy and falling back to the predefined stats only if the computed stats don't exist.

File: helios/train/callbacks/evaluator_callback.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+from helios.evals.datasets import GeobenchDataset
+from helios.evals.embeddings import get_embeddings
+from helios.evals.knn import run_knn
+
+logger = logging.getLogger(__name__)
+
+
+# Geobench classification
+METRIC_NAME = "f1"
```
Comment (gabrieltseng) [inline helios/train/callbacks/evaluator_callback.py]: The metric will actually be top-1 accuracy for eurosat. `run_knn` only returns the f1 score in the multilabel classification case (e.g. BigEarthNet).

See logic here: https://github.com/allenai/helios/blob/8806818a5aae5f9df7559a23e89dc02c94f1f6fe/helios/evals/knn.py#L37

File: helios/data/normalize.py (hunk 1 above)
Comment (yawenzzzz) [reply helios/data/normalize.py]: agree, I'm borrowing the pre-computed mean & std from Presto right now, we will need to compute those stats offline later, and update the numbers. OK, I ended up follow Gabi's suggestion, use computed strategy first and then falling back to predefined stats.

File: helios/nn/flexihelios.py (hunk 2 above)
Comment (yawenzzzz) [reply helios/nn/flexihelios.py:354]: fixed.

File: helios/train/callbacks/evaluator_callback.py (hunk 3 above)
Comment (yawenzzzz) [reply helios/train/callbacks/evaluator_callback.py:88]: These callbacks are tightly integrated with the trainer, so testing is not very straightforward. They also require real evaluators. I tested it with real runs, and it worked.

File: helios/train/loss.py (hunk 4 above)
Comment (yawenzzzz) [reply helios/train/loss.py]: Fixed, thanks!

File: helios/train/train_module/latent_mim.py (hunk 5 above)
Comment (yawenzzzz) [reply helios/train/train_module/latent_mim.py]: I think it’s good to use `record_metric` for logging train loss within `train_batch`, as it’s the only metric we record. We can add a static method later if more metrics are needed.

File: helios/train/callbacks/evaluator_callback.py (hunk 6 above)
Comment (yawenzzzz) [reply helios/train/callbacks/evaluator_callback.py]: Thanks for the clarification, updated the metric name!
